test/common/http/conn_manager_impl_test.cc

```diff
@@ -1451,9 +1459,199 @@ TEST_F(HttpConnectionManagerImplTest, PerStreamIdleTimeoutAfterBidiData) {
+
+  EXPECT_EQ(0U, stats_.named_.downstream_rq_timeout_.value());
+}
+
+TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnCompleteRequestWithHeader) {
+  request_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).Times(1).WillOnce(Invoke([&](Buffer::Instance&) -> void {
+    Event::MockTimer* request_timer = setUpTimer();
+    EXPECT_CALL(*request_timer, enableTimer(request_timeout_)).Times(1);
+
+    StreamDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+    HeaderMapPtr headers{
+        new TestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "GET"}}};
+
+    EXPECT_CALL(*request_timer, disableTimer()).Times(1);
+    decoder->decodeHeaders(std::move(headers), true);
+  }));
+
+  Buffer::OwnedImpl fake_input("1234");
+  conn_manager_->onData(fake_input, false); // kick off request
+
+  EXPECT_EQ(0U, stats_.named_.downstream_rq_timeout_.value());
+}
+
+TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnCompleteRequestWithData) {
+  request_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> void {
+    Event::MockTimer* request_timer = setUpTimer();
+    EXPECT_CALL(*request_timer, enableTimer(request_timeout_)).Times(1);
+
+    StreamDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+    HeaderMapPtr headers{
+        new TestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "POST"}}};
+    decoder->decodeHeaders(std::move(headers), false);
+
+    EXPECT_CALL(*request_timer, disableTimer()).Times(1);
+    decoder->decodeData(data, true);
+  }));
+
+  Buffer::OwnedImpl fake_input("1234");
+  conn_manager_->onData(fake_input, false);
+
+  EXPECT_EQ(0U, stats_.named_.downstream_rq_timeout_.value());
+}
+
+TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnCompleteRequestWithTrailers) {
+  request_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> void {
+    Event::MockTimer* request_timer = setUpTimer();
+    EXPECT_CALL(*request_timer, enableTimer(request_timeout_)).Times(1);
+    StreamDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+
+    HeaderMapPtr headers{
+        new TestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "GET"}}};
+    decoder->decodeHeaders(std::move(headers), false);
+    decoder->decodeData(data, false);
+
+    EXPECT_CALL(*request_timer, disableTimer()).Times(1);
+    HeaderMapPtr trailers{new TestHeaderMapImpl{{"foo", "bar"}}};
+    decoder->decodeTrailers(std::move(trailers));
+  }));
+
+  Buffer::OwnedImpl fake_input("1234");
+  conn_manager_->onData(fake_input, false);
+
+  EXPECT_EQ(0U, stats_.named_.downstream_rq_timeout_.value());
+}
+
+TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnEncodeHeaders) {
```

Comment:
Shouldn't there be be tests here for Encode100Continue and what(ever) its result is supposed to be?

Should there be tests here (honest question; I don't know) for what happens if something other than the request timeout forces a local reply?

Reply:
Auni, did you address this point in some fashion?  (The tests for the behavior; I understand that we don't want to do anything special in the code for 100Continue.)

Reply:
oh oops, why did I miss this comment...Thanks for flagging it, adding that test now.

Reply:
kk added a test to assert encode100 does not disarm